test/test_common/logging.h

```diff
@@ -61,30 +61,104 @@ class LogRecordingSink : public Logger::SinkDelegate {
+      }                                                                                            \
+    }                                                                                              \
+    if (!failed_expectations.empty()) {                                                            \
+      std::string failed_message;                                                                  \
+      absl::StrAppend(&failed_message, "\nLogs:\n ", absl::StrJoin(log_recorder.messages(), " "),  \
+                      "\n Do NOT contain:\n");                                                     \
+      for (const auto& expectation : failed_expectations) {                                        \
+        absl::StrAppend(&failed_message, "  '", expectation.first, "', '", expectation.second,     \
+                        "'\n");                                                                    \
+      }                                                                                            \
+      FAIL() << failed_message;                                                                    \
+    }                                                                                              \
+  } while (false)
+
+// Validates that when stmt is executed, log message containing substr and loglevel will NOT be
+// emitted. Failure message e.g.,
+//
+// Logs:
+//  [2018-04-12 05:51:00.245][7290192][warning][upstream] grpc_mux_impl.cc:160] Received gRPC
+//  [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:63] Called bar
+//  Should NOT contain:
+//   'warning', 'Received gRPC’
+#define EXPECT_LOG_NOT_CONTAINS(loglevel, substr, stmt)                                            \
+  do {                                                                                             \
+    LogLevelSetter save_levels(spdlog::level::trace);                                              \
+    LogRecordingSink log_recorder(Logger::Registry::getSink());                                    \
+    stmt;                                                                                          \
+    for (const std::string& message : log_recorder.messages()) {                                   \
+      if ((message.find(substr) != std::string::npos) &&                                           \
+          (message.find(loglevel) != std::string::npos)) {                                         \
+        FAIL() << "\nLogs:\n " << absl::StrJoin(log_recorder.messages(), " ")                      \
+               << "\n Should NOT contain:\n '" << loglevel << "', '" << substr "'\n";              \
+      }                                                                                            \
+    }                                                                                              \
+  } while (false)
+
+// Validates that when stmt is executed, exactly one log message containing substr will be emitted.
```

Comment:
I don't think this is technically correct, since it will match if many log messages containing substr are emitted as well, right?

Reply:
Nice catch! It should be: `Validates that when stmt is executed, the supplied substring matches at least one log message.`